File: app/src/main/java/net/emilla/math/BitwiseOperator.java
package net.emilla.math;

import java.math.BigInteger;

enum BitwiseOperator implements CalcOperator<BigInteger>, BitwiseToken {
    OR(-3) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.or(b);
        }
    },
    XOR(-2) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.xor(b);
        }
    },
    AND(-1) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.and(b);
        }
    },
    SHIFT_LEFT(0) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.shiftLeft(Maths.exactInt(b));
        }
    },
    SHIFT_RIGHT(0) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.shiftRight(Maths.exactInt(b));
        }
    },
    ADD(1) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.add(b);
        }
    },
    SUBTRACT(1) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.subtract(b);
        }
    },
    MULTIPLY(2) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.multiply(b);
        }
    },
    DIVIDE(2) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.divide(b);
        }
    },
    MOD(2) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.mod(b);
        }
    },
    STUPID_POW(3) {
        @Override
        public BigInteger apply(BigInteger a, BigInteger b) {
            return a.pow(Maths.exactInt(b));
        }
    },
;
    private static final BitwiseOperator[] sValues = values();

    private final int mPrecedence;

    BitwiseOperator(int precedence) {
        mPrecedence = precedence;
    }

    public static BitwiseOperator of(int ordinal) {
        return sValues[ordinal];
    }

    public static BitwiseOperator of(char token) {
        return switch (token) {
            case '|' -> OR;
            case '^' -> XOR;
            case '&' -> AND;
            case '<' -> SHIFT_LEFT;
            case '>' -> SHIFT_RIGHT;
            case '+' -> ADD;
            case '-' -> SUBTRACT;
            case '*' -> MULTIPLY;
            case '/' -> DIVIDE;
            case '%' -> MOD;
            default -> throw new IllegalArgumentException("Invalid bitwise operator");
        };
    }

    @Override
    public int precedence() {
        return mPrecedence;
    }

    @Override
    public final boolean isRightAssociative() {
        return this == STUPID_POW;
    }
}

File: app/src/main/java/net/emilla/math/BitwiseOperator.kt
package net.emilla.math

import net.emilla.math.CalcToken.BitwiseToken
import kotlin.math.pow

internal enum class BitwiseOperator(
    @JvmField val precedence: Int,
    @JvmField val rightAssociative: Boolean
) : BitwiseToken {
    OR(-3, false) {
        override fun Long.apply(n: Long) = this or n
    },
    XOR(-2, false) {
        override fun Long.apply(n: Long) = this xor n
    },
    AND(-1, false) {
        override fun Long.apply(n: Long) = this and n
    },
    SHL(0, false) {
        override fun Long.apply(n: Long) = this shl n.toInt()
    },
    SHR(0, false) {
        override fun Long.apply(n: Long) = this shr n.toInt()
    },
    USHR(0, false) {
        override fun Long.apply(n: Long) = this ushr n.toInt()
    },
    PLUS(1, false) {
        override fun Long.apply(n: Long) = this + n
    },
    MINUS(1, false) {
        override fun Long.apply(n: Long) = this - n
    },
    TIMES(2, false) {
        override fun Long.apply(n: Long) = this * n
    },
    DIV(2, false) {
        override fun Long.apply(n: Long) = this / n
    },
    MOD(2, false) {
        override fun Long.apply(n: Long) = this % n
    },
    POW(3, true) {
        override fun Long.apply(n: Long) = toDouble().pow(n.toDouble()).toLong()
    };

    abstract fun Long.apply(n: Long): Long

    companion object {
        @JvmField
        val LPAREN: BitwiseOperator? = null

        @JvmStatic
        fun of(token: String) = when (token) {
            // todo: natural language like "add", "to the power of", ..
            "|" -> OR
            "^" -> XOR
            "&" -> AND
            "<<" -> SHL
            ">>" -> SHR
            ">>>" -> USHR
            "+" -> PLUS
            "-" -> MINUS
            "*" -> TIMES
            "/" -> DIV
            "%" -> MOD
            else -> throw IllegalArgumentException()
        }
    }
}
